src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -143,7 +144,11 @@ public ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = d
         {
             if (_pipeWriterCompleted)
             {
-                return default;
+                return new ValueTask<FlushResult>(new FlushResult(false, true));
+            }
+            if (_pendingFlushCanceled)
+            {
+                return new ValueTask<FlushResult>(new FlushResult(true, false));
```

Comment:
Okay, so there is an issue with this. `CancelPendingFlush()` (and `CancelPendingRead()` FWIW) is not terminal. After returning the first `FlushResult`, it should be possible to flush again normally without it being immediately canceled.

Also, `CancelPendingFlush()` can be called before or during an ongoing flush. It should not prevent a flush from happening even if it's called first. It should just cause the flush to immediately return with a cancelled `FlushResult`. Then next call to `FlushAsync()` shouldn't return a cancelled `FlushResult` unless `CancelPendingFlush()` was called again.

I think this part should have been taken care of already by the call to `_pipeWriter.CancelPendingFlush();` Is it possible this was already behaving correctly, but the intended behavior wasn't well understood?

Reply:
So i think the problem is that PipeFlusher either forgets that it has been cancelled or doesn't return correctly. I will look into that, converting this PR to draft for now